### PR TITLE
Use `aerich` for database migrations

### LINKED ISSUE #1107: Database migrations with Aerich
https://github.com/tortoise/aerich

### PR DESCRIPTION
**Pull request by bigherc18**
This PR is about using `aerich` to add support for database migrations

Here's a demo: https://asciinema.org/a/Sw7OwQZKmWyznV3xgohjd59OT

I added a new config called `migrations_dir` to specify where migration files should be stored

```yaml [dipdup.yaml]
database:
  migrations_dir: ./migrations
```

A a set of commands have been added to manage database migrations:

- `dipdup schema history` to show the migration history
- `dipdup schema heads` to show the show the current available heads
- `dipdup schema migrate` to create a new migration changes file
- `dipdup schema upgrade` to upgrade the database to the latest or a specified version
- `dipdup schema downgrade` to downgrade the database to a previous version

Theses commands invoke `aerich`'s commands under the hood

The migrations directory should be initialized before using any of these commands, either with `dipdup schema init` or by running the indexer with `dipdup run`.

Note that `aerich` commands can't be used directly, as it expects the config to be stored in `pyproject.toml`

### CODE REVIEW

File: src/dipdup/cli.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
 
 import click
 import uvloop
+from aerich.cli import cli as aerich_cli
```
Comment (droserasprout) [inline src/dipdup/cli.py:20]: This import should be lazy, since it takes 0.6s and  `dipdup.cli` module is imported for every command.

Comment (droserasprout) [inline src/dipdup/cli.py]: This import should be lazy, since it takes 0.6s and dipdup.cli module is imported for every command.

File: src/dipdup/cli.py [hunk 2]
```diff
@@ -562,7 +580,52 @@ async def hasura_configure(ctx: click.Context, force: bool) -> None:
 @_cli_wrapper
 async def schema(ctx: click.Context) -> None:
     """Commands to manage database schema."""
-    pass
+    from pathlib import Path
+
+    from dipdup.aerich import create_aerich_command
```
Comment (droserasprout) [inline src/dipdup/cli.py]: This import is not lazy enough. You could do it right before the call.

File: src/dipdup/config/__init__.py [hunk 3]
```diff
@@ -98,11 +99,13 @@ class SqliteDatabaseConfig:
     :param kind: always 'sqlite'
     :param path: Path to .sqlite file, leave default for in-memory database (`:memory:`)
     :param immune_tables: List of tables to preserve during reindexing
+    :param migrations_dir: Where to store migrations files (relative to project root)
```
Comment (droserasprout) [inline src/dipdup/config/__init__.py]: I suggest removing an ability to configure this path and hardcode it in `DipDupPackage` class instead (but don't create by default). We already use fixed directory paths for every other section; it makes things much simpler to maintain.

File: pyproject.toml [hunk 4]
```diff
@@ -80,6 +80,7 @@ dependencies = [
     "tortoise-orm==0.21.6",
     "uvloop~=0.20",
     "web3~=7.2",
+    "aerich~=0.7.2",
```
Comment (droserasprout) [inline pyproject.toml]: Let's make this dependency optional. A majority of users don't need migrations, and dependency tree is already very bloated. If it's not installed -- just ignore the whole migrations part.

File: src/dipdup/cli.py [hunk 5]
```diff
@@ -562,7 +580,52 @@ async def hasura_configure(ctx: click.Context, force: bool) -> None:
...
+    return click.Command(
+        name=command.name,
+        callback=wrapper,
+        params=command.params,
+        help=command.help,
+        short_help=command.short_help,
+    )
+
+
+schema.add_command(aerich_cli.commands['history'])
```
Comment (droserasprout) [inline src/dipdup/cli.py]: Please wrap this with try except to silently fail when aerich is not installed.

**Review by droserasprout — CHANGES_REQUESTED** [review]
Hi! Thanks for your contribution! It's a great code with many smart solutions; I like every piece of it! My only concern is making this integration optional in every way: dependencies, imports, files in the project package.

**Comment by bigherc18** [comment]
Do you guys prefer `rebase` or `merge` with `next` to fix the conflicts ?

**Comment by droserasprout** [comment]
@bigherc18 We squash every merge request, so it's up to you what's inside.

**Comment by bigherc18** [comment]
Some concerns and limitations I have:

- `aerich` will be confused if the user changes the database. If we want to support changing the database, we have to make aerich aware of which database is being used.
- If you answer "No" to the question "Downgrade is dangerous, which may cause data loss. Are you sure? [y/N]" after running `dipdup schema downgrade`, the CLI hangs there forever.
- `aerich` has confusing messages after running `heads` and `history`. It makes the user feel like they have to take action when everything is actually fine: "No available heads, try migrate first."

File: .vscode/settings.json [hunk 6]
```diff
@@ -7,5 +7,8 @@
     },
     "files.associations": {
         "*.env.default": "shellscript"
-    }
-}
\ No newline at end of file
+    },
```
Comment (droserasprout) [inline .vscode/settings.json]: Please, remove this part.

File: src/dipdup/database.py [hunk 7]
```diff
@@ -69,11 +69,14 @@ async def tortoise_wrapper(
 
     model_modules: dict[str, Iterable[str | ModuleType]] = {
         'int_models': ['dipdup.models'],
+        'models': ['dipdup.models'],
```
Comment (droserasprout) [inline src/dipdup/database.py]: This line is a duplicate. Without it mypy seems to be okay with `model_modules['models'] = [models]`

File: src/dipdup/cli.py [hunk 8]
```diff
@@ -562,7 +579,59 @@ async def hasura_configure(ctx: click.Context, force: bool) -> None:
...
+    config: DipDupConfig = ctx.obj.config
+
+    if ctx.invoked_subcommand in AERICH_CMDS:
+        from dipdup.package import DipDupPackage
+
+        migrations_dir = DipDupPackage(config.package_path).migrations
+
+        if not migrations_dir.exists():
+            echo(
+                f"""Database migrations is not initialize at {migrations_dir}.
```
Comment (droserasprout) [inline src/dipdup/cli.py]: *are not initialized

File: src/dipdup/cli.py [hunk 9]
```diff
@@ -562,7 +579,59 @@ async def hasura_configure(ctx: click.Context, force: bool) -> None:
...
+                 Run `dipdup schema init` or just run `dipdup run` and it'll be initialized automatically."""
+            )
+            raise click.Abort()
```
Comment (droserasprout) [inline src/dipdup/cli.py]: Please, raise `dipdup.exceptions.ProjectPackageError` here so we get a nicer help message.

**Review by droserasprout — CHANGES_REQUESTED** [review]
Your last comment shows such a great amount of testing done! 🥰 I think, there's nothing critical. Yes, running commands in random order on multiple databases sometimes confuse aerich, but all real cases works perfectly.

I have just a few nitpicks. Then make sure that `make all docs` command passes, add yourself to the `thanks.md` file, and let's merge it!

**Comment by bigherc18** [comment]
> I think, there's nothing critical. Yes, running commands in random order on multiple databases sometimes confuse aerich, but all real cases works perfectly.

Thank you for your feedback but I just want to make sure I explained my concerns clear enough for you

1. Multiple databases not supported by migrations

TLDR; if a user uses multiple databases at the same time, migrations is not supported

I see dipdup makes it easy for users to swap databases by just combining different config files with multiple `-c` args, the problem is: if a user swaps his database after doing some migrations in the first one, `aerich` is now really confused and I don't even know what will happen, the `aerich` table will be empty (or contains migrations history for the second database) but the `migrations` folder still contains files from the first database

I think this worth at least a note in the docs 

3. CLI handing

Here's a little video demonstrating the issue:

demo

**Review by droserasprout — APPROVED** [review]
